packages/fetchai/connections/p2p_libp2p/libp2p_node.go

```diff
@@ -0,0 +1,414 @@
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * This program demonstrate a simple chat application using p2p communication.
+ *
+ */
+package main
+
+import (
+	"bufio"
+	"context"
+	"encoding/binary"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"os/signal"
+	"sync"
+
+	"github.com/libp2p/go-libp2p"
+	"github.com/libp2p/go-libp2p-core/crypto"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/peerstore"
+
+	ds "github.com/ipfs/go-datastore"
+	dsync "github.com/ipfs/go-datastore/sync"
+	dht "github.com/libp2p/go-libp2p-kad-dht"
+	rhost "github.com/libp2p/go-libp2p/p2p/host/routed"
+	"github.com/multiformats/go-multiaddr"
+
+	aea "libp2p_node/aea"
+
+	btcec "github.com/btcsuite/btcd/btcec"
+
+	proto "github.com/golang/protobuf/proto"
+)
+
+// panics if err is not nil
+func check(err error) {
+	if err != nil {
+		panic(err)
+	}
+}
+
+func main() {
+
+	var err error
+
+	// Initialize connection to aea
+	agent := aea.AeaApi{}
+	check(agent.Init())
+	log.Println("successfully initialised API to AEA!")
+
+	// Get node configuration
+
+	// node address (ip and port)
+	// TOFIX(LR) ignoring the host(ip)
+	_, nodePort := agent.Address()
+
+	// node private key
+	key := agent.PrivateKey()
+	prvKey, _, err := KeyPairFromFetchAIKey(key)
+	check(err)
+
+	// entry peers
+	entryPeers := agent.EntryPeers()
+	bootstrapPeers, err := GetPeersAddrInfo(entryPeers)
+	check(err)
+	log.Println(bootstrapPeers)
+
+	// Configure node's multiaddr
+	// TOFIX(LR) binding only to 127.0.0.1 will prevent the peer from being found
+	//   on the dht, when using ipfs bootstrap peers
+	nodeMultiAddr, _ := multiaddr.NewMultiaddr(fmt.Sprintf("/ip4/0.0.0.0/tcp/%d", nodePort))
+
+	// Make a host that listens on the given multiaddress
+	rhost, _, err := setupRoutedHost(nodeMultiAddr, prvKey, bootstrapPeers)
+	log.Println("successfully created libp2p node!")
+
+	// Set a stream handler on host
+	rhost.SetStreamHandler("/aea/1.0.0", func(s network.Stream) {
+		handleAeaStream(s, agent)
+	})
+
+	// Connect to the agent
+	check(agent.Connect())
+	log.Println("successfully connected to AEA!")
+
+	// Receive envelopes from agent and forward to peer
+	go func() {
+		for envel := range agent.Queue() {
+			log.Println("INFO Received envelope from agent:", envel)
```

Comment:
Slightly clarify this, maybe `"INFO Got envelope from AEA and sending to peer:"`